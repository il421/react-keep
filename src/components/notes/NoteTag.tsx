import React from "react";
import { Store, TagsStoreState } from "../../store/store.types";
import { connect } from "react-redux";
interface NoteTagProps {
  tagId: string;
}

interface StateProps {
  tags: TagsStoreState[];
}

type Props = NoteTagProps & StateProps;

const NoteTag: React.FunctionComponent<Props> = ({
  tagId,
  tags,
}): JSX.Element => {
  const getTagById = (tagId: string): string | undefined => {
    const tag = tags.find((t) => t.id === tagId);
    return tag ? tag.value : undefined;
  };

  return <div className="note__tag">{getTagById(tagId)}</div>;
};

const mapStateToProps = (state: Store): StateProps => {
  return {
    tags: state.tags,
  };
};

export default connect<StateProps, {}, NoteTagProps, Store>(mapStateToProps)(
  NoteTag
);
