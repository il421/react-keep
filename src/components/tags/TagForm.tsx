import React from "react";
import { Store, TagsStoreState } from "../../store/store.types";
import { ThunkDispatch } from "redux-thunk";
import { connect } from "react-redux";
import { TextInputField, BaseForm } from "../form";
import { Placeholders, nameOf, Errors } from "../../common";
import { handleAddTag } from "../../actions/tags";
import "../../styles/components/tags/_tags-form.scss";
import { IconButton } from "../ui-components";
import { ValidationErrors } from "final-form";

interface TagFormProps {}

interface StateProps {
  tags: TagsStoreState[];
}

interface DispatchProps {
  addTag: (value: string) => void;
}
type Props = TagFormProps & DispatchProps & StateProps;

interface TagFormValues {
  value: string | undefined;
}

const getValidationErrors = (values: TagFormValues): ValidationErrors => {
  const errors: ValidationErrors = {};
  const { value } = values;

  if (value && value.length === 0) {
    errors.value = Errors.required;
  } else if (value && value.length > 20) {
    errors.value = Errors.tag;
  }
  return errors;
};

class TagForm extends React.PureComponent<Props> {
  private nameOf = nameOf<TagFormValues>();

  render() {
    return (
      <BaseForm<TagFormValues>
        initialValues={{ value: undefined }}
        onSubmit={(values: TagFormValues) =>
          values.value && this.props.addTag(values.value)
        }
        onCancel={() => {}}
        classNames={{
          form: "tags-form",
        }}
        validate={getValidationErrors}
        resetAfterSubmitting={true}
        getButtons={(isDisable: boolean) => (
          <IconButton
            className="tags-form__submit"
            icon="plus-circle"
            size="1x"
            type="submit"
            disabled={isDisable}
          />
        )}
      >
        <TextInputField
          name={this.nameOf("value")}
          type="text"
          placeholder={Placeholders.newTag}
          className="tags-form__field"
        />
      </BaseForm>
    );
  }
}

const mapStateToProps = (state: Store): StateProps => {
  return {
    tags: state.tags,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  addTag: (value: string) => dispatch(handleAddTag(value)),
});

export default connect<StateProps, DispatchProps, TagFormProps, Store>(
  mapStateToProps,
  mapDispatchToProps
)(TagForm);
