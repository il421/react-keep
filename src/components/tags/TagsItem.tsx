import React, { FunctionComponent } from "react";

import { AlignItems, JustifyContent } from "../../common";
import { Tag } from "../../store/store.types";
import "../../styles/components/tags/_tags-item.scss";
import { FlexBox, IconButton } from "../ui-components";

export interface TagsItemProps {
  tag: Tag;
  removeTag: (id: string) => void;
  checked: boolean;
  onChange: (id: string) => void;
}

export const TagsItem: FunctionComponent<TagsItemProps> = ({
  tag,
  removeTag,
  checked,
  onChange
}) => {
  return (
    <FlexBox
      justifyContent={JustifyContent.spaceBetween}
      alignItems={AlignItems.center}
      className="tags-item"
    >
      <FlexBox justifyContent={JustifyContent.start}>
        <div className="tags-item__checkbox">
          <input
            ref={`${tag.id}-list`}
            id={`${tag.id}-list`}
            type="checkbox"
            checked={checked}
            onChange={() => onChange(tag.id)}
          />
          <label htmlFor={`${tag.id}-list`} />
        </div>
        <div className="tags-item__value">{tag.value}</div>
      </FlexBox>

      <IconButton icon="times" onButtonClick={() => removeTag(tag.id)} />
    </FlexBox>
  );
};
