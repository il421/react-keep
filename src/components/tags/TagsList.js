import React from 'react';
import { connect } from 'react-redux';
import xor from 'lodash/xor';

import { setTagsFilter } from '../../actions/filters';

export class TagsList extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      tagFilters: []
    };
  }

  handleTagsCheck = async (evt) => {
    const value = evt.target.value;

    // toggle elements of array with xor lodash
    await this.setState((prevState) => ({
      tagFilters: xor(prevState.tagFilters, [value])
    }));

    this.props.setTagsFilter(this.state.tagFilters);
  }

  render() {
    return (
      <div className="tags-list">
        <h2 className="tags-list__title">Filter by Tag</h2>
        <div className="tags-list__filters">
          {
            this.props.userTags.length > 0 && (
              this.props.userTags.map((tag, index) => (
                <div key={ index } className="tags-list__filter">
                  <input
                    ref={ tag.id + 'list' }
                    id={ tag.id + 'list' }
                    type="checkbox"
                    value={ tag.id }
                    defaultChecked={ true }
                    onChange={ this.handleTagsCheck }
                  />
                  <label htmlFor={ tag.id + 'list' } className={ `tag-${index}` }/>
                  <div>{ tag.value }</div>
                </div>
              ))
            )
          }
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setTagsFilter: (tagFilters) => dispatch(setTagsFilter(tagFilters))
  };
};

const mapStateToProps = (state) => {
  return {
    userTags: state.tags.list
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TagsList);
