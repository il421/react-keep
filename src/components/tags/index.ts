export * from "./TagForm";
export * from "./Tags";
export * from "./TagsItem";
export * from "./TagsList";
