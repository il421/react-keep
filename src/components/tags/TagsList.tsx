import React from "react";
import { FlexBox } from "../ui-components";
import { JustifyContent } from "../../common";
import {
  FiltersStoreState,
  Store,
  TagsStoreState,
} from "../../store/store.types";
import { ThunkDispatch } from "redux-thunk";
import { connect } from "react-redux";
import TagsItem from "./TagsItem";
import { handleRemoveTag } from "../../actions/tags";
import { setTagsFilter } from "../../actions/filters";
import { removeTagFromNotes } from "../../actions/notes";
import "../../styles/components/tags/_tags-list.scss";

interface TagsListProps {}

interface StateProps {
  tags: TagsStoreState[];
  filters: FiltersStoreState;
}

interface DispatchProps {
  removeTag: (id: string) => void;
  toggleTag: (id: string) => void;
  removeTagFromNotes: (id: string) => void;
}

export type Props = DispatchProps & StateProps;

export class TagsListBase extends React.PureComponent<Props> {
  render() {
    return (
      <FlexBox
        vertical
        justifyContent={JustifyContent.start}
        className="tags-list"
      >
        {this.props.tags.map((tag) => (
          <TagsItem
            key={tag.id}
            tag={tag}
            removeTag={(id: string) => {
              this.props.removeTag(id);
              this.props.removeTagFromNotes(id);
            }}
            checked={this.props.filters.tagFilters.includes(tag.id)}
            onChange={this.props.toggleTag}
          />
        ))}
      </FlexBox>
    );
  }
}

const mapStateToProps = (state: Store): StateProps => {
  return {
    tags: state.tags,
    filters: state.filters,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  removeTag: (id: string) => dispatch(handleRemoveTag(id)),
  toggleTag: (id: string) => dispatch(setTagsFilter(id)),
  removeTagFromNotes: (id: string) => dispatch(removeTagFromNotes(id)),
});

export const TagsList = connect<
  StateProps,
  DispatchProps,
  TagsListProps,
  Store
>(
  mapStateToProps,
  mapDispatchToProps
)(TagsListBase);
