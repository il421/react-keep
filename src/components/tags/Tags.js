import React, { Component, createRef } from 'react';
import {connect} from 'react-redux';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Modal from 'react-modal';

import { handleAddTag, handleRemoveTag, handleUpdateTag } from '../../actions/tags';
import { removeTagFromNotes, updateNotesTag } from '../../actions/notes';

import TagsList from './TagsList';

export class Tags extends Component{
  constructor(props) {
    super(props);
    this.state = {
      displayEditTags: false,
      editableTag: null,
      tag: '',
    };

    // input new tag ref
    this.refNewInput = createRef();
  }

  toggleEditTags = () => {
    this.setState((prevState) => ({
      displayEditTags: !prevState.displayEditTags
    }));
    this.props.hideSidebar();
  }

  onTagChange = (evt) => {
    const value = evt.target.value;
    this.setState(() => ({
      tag: value.trim()
    }));
  }

  onCurrentTagUpdate = (id, update) => {
    if (update.trim().length > 0) {
      this.props.handleUpdateTag(id, update);
      this.props.updateNotesTag(id, update);
    }

    this.setEditableTag(id);
  }

  addNewTag = () => {
    if (this.state.tag !== '') {
      this.props.handleAddTag(this.state.tag);
    } else {
      // Add toast
    }

    // clean state.tag and local input value by ref
    this.refNewInput.current.value = '';
    this.setState({ tag: '' });
  }

  removeTag = (id) => {
    this.props.handleRemoveTag(id);
    this.props.removeTagFromNotes(id);
  }

  setEditableTag = (id) => {
    let editableTag = null;

    if (this.state.editableTag === null) {
      editableTag = id;
    }

    this.setState({ editableTag });
  }

  render() {
    return (
      <div className="tags">

        <div className="tags__filters">
          <TagsList userTags={ this.props.tags } name={ 'list' } />
        </div>

        <div className="tags__add-edit pointer">
          <FontAwesomeIcon icon="edit" size="2x" />
          <button onClick={ this.toggleEditTags } className="button--link">Add/Edit Tags</button>
        </div>

        <Modal
          isOpen={ this.state.displayEditTags }
          onRequestClose={ this.toggleEditTags }
          closeTimeoutMS={ 200 }
          className="tags__modal"
          ariaHideApp={ false }
        >
          <div>
            <div className="tags__title">Add/Edit Tags</div>
            <div className="tags__add">
              <input
                id="newTag"
                ref={ this.refNewInput }
                type="text"
                placeholder="Type a new tag ..."
                onChange={ this.onTagChange }
                onKeyPress={ (evt) => evt.key === 'Enter' && this.addNewTag() }
                disabled={ this.state.editableTag !== null }
              />
              <button className="button--link" onClick={ this.addNewTag }>+</button>
            </div>

            <div className="tags__list">
              {
                this.props.tags.map((tag, index) => (
                  <div className="tags__item item" key={ index }>

                    {
                      this.state.editableTag === tag.id ?
                      // if id is in editableTags array display input to edit the tag
                        (
                          <>
                            <input
                              id="editableTag"
                              type="text"
                              autoFocus={ true }
                              onChange={ this.onTagChange }
                              onKeyPress={ (evt) => evt.key === 'Enter' && this.onCurrentTagUpdate(tag.id, this.state.tag) }
                              placeholder={ tag.value }
                            />
                            <div className="item__actions">
                              <div className="pointer" onClick={ () => this.onCurrentTagUpdate(tag.id, this.state.tag) }>
                                <FontAwesomeIcon icon="edit" size="1x" />
                              </div>
                              <div className="pointer" onClick={ () => this.setEditableTag(tag.id) }>
                                <FontAwesomeIcon icon="times" size="1x" />
                              </div>
                            </div>
                          </>
                        ) : (
                          <>
                            <div className="pointer" onClick={ () => this.setEditableTag(tag.id) }>
                              <FontAwesomeIcon icon="edit" size="1x" />
                            </div>
                            <div className="tags__name">{ tag.value }</div>
                            <div className="pointer" onClick={ () => {this.removeTag(tag.id)} }>
                              <FontAwesomeIcon icon="times" size="1x" />
                            </div>
                          </>
                        )
                    }
                  </div>
                ))
              }
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  handleAddTag: (tag) => dispatch(handleAddTag(tag)),
  handleRemoveTag: (id) => dispatch(handleRemoveTag(id)),
  removeTagFromNotes: (id) => dispatch(removeTagFromNotes(id)),
  handleUpdateTag: (id, update) => dispatch(handleUpdateTag(id, update)),
  updateNotesTag: (id, update) => dispatch(updateNotesTag(id, update))
}
);

const mapStateToProps = (state) => {
  return {
    tags: state.tags
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Tags);
