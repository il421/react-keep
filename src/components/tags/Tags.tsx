import React from "react";
import "../../styles/components/tags/_tags.scss";
import TagForm from "./TagForm";
import TagsList from "./TagsList";

interface TagsProps {}

class Tags extends React.PureComponent<TagsProps> {
  render() {
    return (
      <div className="tags">
        <TagForm />
        <TagsList />
      </div>
    );
  }
}

export default Tags;
