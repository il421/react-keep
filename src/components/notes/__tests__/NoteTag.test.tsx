import React from "react";
import { shallow } from "enzyme";
import { NoteTagBase } from "../NoteTag";
import toJson from "enzyme-to-json";
import { tags } from "../../../testData/tags";

test("should render NoteTag", () => {
  const textWrapper = shallow(<NoteTagBase tagId={"1"} tags={tags} />);
  expect(toJson(textWrapper)).toMatchSnapshot();
});
