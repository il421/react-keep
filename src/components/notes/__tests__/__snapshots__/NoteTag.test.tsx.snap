// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NoteTag 1`] = `
<div
  className="note__tag"
>
  Tag1
</div>
`;
